vm/vm.go

```diff
@@ -846,13 +854,28 @@ func (vm *VM) BuildBlock(ctx context.Context) (snowman.Block, error) {
 		vm.snowCtx.Log.Warn("unable to get preferred block", zap.Error(err))
 		return nil, err
 	}
-	blk, err := chain.BuildBlock(ctx, vm, preferredBlk)
+	preferredView, err := preferredBlk.View(ctx, true)
+	if err != nil {
+		vm.snowCtx.Log.Warn("unable to get preferred block view", zap.Error(err))
+		return nil, err
+	}
+	executionBlk, executedBlk, view, err := vm.chain.BuildBlock(ctx, preferredView, preferredBlk.ExecutionBlock)
 	if err != nil {
 		// This is a DEBUG log because BuildBlock may fail before
 		// the min build gap (especially when there are no transactions).
 		vm.snowCtx.Log.Debug("BuildBlock failed", zap.Error(err))
 		return nil, err
 	}
+	blk := &StatefulBlock{
```

Comment:
TODO: centralize where we create stateful block instances to make it easier to reason about